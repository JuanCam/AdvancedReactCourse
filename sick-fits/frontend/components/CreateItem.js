import React from 'react';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

const CREATE_ITEM_MUTATION = gql`
mutation CREATE_ITEM_MUTATION(
        $title: String!
        $description: String!
        $image: String
        $largeImage: String
        $price: Int!
    ) {
    createItem(
        title: $title
        description: $description
        image: $image
        largeImage: $largeImage
        price: $price
    ) {
        id
    }
}
`;

class CreateItem extends React.Component {
    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0,
    };

    handleChange = ({target}) => {
        const {name, type, value} = target;
        const val = (type === 'number')
        ? parseFloat(value) : value;
        this.setState({[name]: val});
    }

    uploadFile = async ({target}) => {
        const files = target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'sickfits');

        const res = await fetch('https://api.cloudinary.com/v1_1/ddxx7sbgs/image/upload', 
        {
            method: 'POST',
            body: data
        });
        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage:  file.eager[0].secure_url
        });
    }

    render() {
        return (
            <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
             {(createItem, {loading, error, called, data}) => (
                    <Form onSubmit={async (e) => {
                        // Stop the form from submiting
                        e.preventDefault();
                        // Call mutation
                        const res = await createItem();
                        // Go to the single item page
                        Router.push({
                            pathname: '/item',
                            query: {id: res.data.createItem.id}
                        })

                    }}>
                    <Error error={error} />
                    
                    <fieldset disabled={loading} aria-bussy={loading}>

                        <label htmlFor="file">Image File</label>
                        <input type="file"
                            id="file"
                            name="file"
                            placeholder="Upload file"
                            required
                            onChange={this.uploadFile} />
                        {this.state.image && <img src={this.state.image} width="150" alt="Upload preview"></img>}

                        <label htmlFor="title">Title</label>
                        <input type="text" 
                        id="title"
                        name="title"
                        placeholder="Title" 
                        required 
                        value={this.state.title}
                        onChange={this.handleChange}/>

                        <label htmlFor="price">Price</label>
                        <input type="number"
                            id="price"
                            name="price"
                            placeholder="Price"
                            required
                            value={this.state.price}
                            onChange={this.handleChange} />

                        <label htmlFor="description">Description</label>
                        <textarea
                            id="description"
                            name="description"
                            placeholder="Enter a description"
                            required
                            value={this.state.description}
                            onChange={this.handleChange} />
                        <button type="submit">Submit</button>
                    </fieldset>
                    </Form>
                )}
            </Mutation>)
    }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };