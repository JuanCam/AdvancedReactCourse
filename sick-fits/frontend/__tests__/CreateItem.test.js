import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';
import Router from 'next/router';

const mockImageUrl = 'https://dog.com/dog.jpg';

// Mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
    json: () => ({
        secure_url: mockImageUrl,
        eager: [{ secure_url: mockImageUrl }]
    })
});

describe('<CreateItem />', () => {
    it('renders and matches snapshot', () => {
        const wrapper = mount(<MockedProvider>
            <CreateItem />
        </MockedProvider>);
        const form = wrapper.find('form');
        expect(toJSON(form)).toMatchSnapshot();
    });

    it('uploads a file when changed', async () => {

        const wrapper = mount(<MockedProvider>
            <CreateItem />
        </MockedProvider>);

        const input = wrapper.find('input[type="file"]');
        input.simulate('change', { target: { files: [mockImageUrl] } });
        await wait();
        const component = wrapper.find('CreateItem').instance();
        expect(component.state.image).toEqual(mockImageUrl);
        expect(component.state.largeImage).toEqual(mockImageUrl);
        expect(global.fetch).toHaveBeenCalled();
        global.fetch.mockReset();
    });

    it('hanldes state updating', async () => {

        const wrapper = mount(<MockedProvider>
            <CreateItem />
        </MockedProvider>);

        wrapper.find('#title').simulate('change', { target: { value: 'testing', name: 'title' } });
        wrapper.find('#price').simulate('change', { target: { value: 50000, name: 'price' } });
        wrapper.find('#description').simulate('change', { target: { value: 'This is the description', name: 'description' } });

        expect(wrapper.find('CreateItem').instance().state)
            .toMatchObject({ 
                title: 'testing',
                price: 50000,
                description: 'This is the description' 
            });

    });

    it('creates the item when the form is submitted', async () => {
        const item= fakeItem();
        const { description, title, price } = item;
        const mocks = [
            {
                request: {
                    query: CREATE_ITEM_MUTATION,
                    variables: {
                        title,
                        description,
                        price,
                        image: '',
                        largeImage: '',
                    }
                },
                result: {
                    data: {
                        createItem: {
                            ...item
                        }
                    }
                }
            }
        ];


        const wrapper = mount(<MockedProvider mocks={mocks}>
            <CreateItem />
        </MockedProvider>);

        wrapper.find('#title').simulate('change', { target: { value: title, name: 'title' } });
        wrapper.find('#price').simulate('change', { target: { value: price, name: 'price' } });
        wrapper.find('#description').simulate('change', { target: { value: description, name: 'description' } });

        Router.router = { push: jest.fn() };
        wrapper.find('form').simulate('submit');

        await wait(50);
        wrapper.update();

        expect(Router.router.push).toHaveBeenCalledWith({ pathname:'/item', query: {id: item.id} });
    });
});
