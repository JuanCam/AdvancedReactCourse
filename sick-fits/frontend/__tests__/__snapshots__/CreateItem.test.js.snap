// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image File
    </label>
    <input
      id="file"
      name="file"
      onChange={[Function]}
      placeholder="Upload file"
      required={true}
      type="file"
    />
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      id="title"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      required={true}
      type="text"
      value=""
    />
    <label
      htmlFor="price"
    >
      Price
    </label>
    <input
      id="price"
      name="price"
      onChange={[Function]}
      placeholder="Price"
      required={true}
      type="number"
      value={0}
    />
    <label
      htmlFor="description"
    >
      Description
    </label>
    <textarea
      id="description"
      name="description"
      onChange={[Function]}
      placeholder="Enter a description"
      required={true}
      value=""
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
